geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/process/JoinProcess.scala

```diff
@@ -0,0 +1,121 @@
+              unique: SimpleFeatureCollection,
+              @DescribeParameter(name = "join", description = "Joined features")
+              join: SimpleFeatureCollection,
+              @DescribeParameter(name = "attribute", description = "Attribute field to join on", min = 1)
+              attribute: String,
+              @DescribeParameter(name = "bins", description = "Return BIN records or regular records", min = 0)
+              bins: java.lang.Boolean,
+              @DescribeParameter(name = "binDtg", description = "Date field to use for BIN records", min = 0)
+              binDtg: String,
+              @DescribeParameter(name = "binTrackId", description = "Track field to use for BIN records", min = 0)
+              binTrackId: String,
+              @DescribeParameter(name = "binLabel", description = "Label field to use for BIN records", min = 0)
+              binLabel: String,
+              monitor: ProgressListener): SimpleFeatureCollection = {
+
+    import org.locationtech.geomesa.filter.ff
+
+    import scala.collection.JavaConversions._
+
+    logger.trace(s"Attempting join query on ${join.getClass.getName}")
+
+    if (join.isInstanceOf[ReTypingFeatureCollection]) {
+      logger.warn("WARNING: layer name in geoserver must match feature type name in geomesa")
+    }
+
+    val descriptor = join.getSchema.getDescriptor(attribute)
+    require(descriptor != null, s"Attribute $attribute does not exist in the joined feature collection")
+
+    require(unique.getSchema.getTypeName == UniqueProcess.SftName &&
+        unique.getSchema.getDescriptor(0).getLocalName == UniqueProcess.AttributeValue,
+      s"Input feature collection is not the result of a unique process. Expected " +
+          s"${SimpleFeatureTypes.encodeType(UniqueProcess.createUniqueSft(descriptor.getType.getBinding, histogram = false))} " +
+          s"but got ${SimpleFeatureTypes.encodeType(unique.getSchema)}"
+    )
+
+    val attributeProperty = ff.property(attribute)
+    val joinFilters = SelfClosingIterator(unique).map { sf =>
+      ff.equals(attributeProperty, ff.literal(sf.getAttribute(0)))
+    }
+
+    val binOptions = if (bins != null && bins) {
+      import org.locationtech.geomesa.utils.geotools.RichSimpleFeatureType.RichSimpleFeatureType
+      val sft = join.getSchema
+      val dtg = Option(binDtg).orElse(sft.getDtgField).getOrElse {
+        throw new IllegalArgumentException("Please specify binDtg for BIN output format")
+      }
+      val track = Option(binTrackId).orElse(sft.getBinTrackId)
+      Some(EncodingOptions(dtg, track, Option(binLabel)))
+    } else {
+      None
+    }
+
+    val result = if (joinFilters.isEmpty) {
+      new DefaultFeatureCollection(null, join.getSchema)
+    } else {
+      val filter = ff.or(joinFilters.toList)
+      val visitor = new QueryVisitor(join, filter)
+      join.accepts(visitor, new NullProgressListener)
```

Comment:
Just to make sure I understand...  If there are a million unique join-field values, then the filter will consist of a million ```equals``` filters all joined into a single ```OR```, and that massive filter will be applied to every feature in the ```join``` collection?  Is that right?